repl.lisp

```diff
@@ -150,39 +151,119 @@
     (sb-int:compiled-program-error (err) (format t "~a~%" err))
     (undefined-function (fun) (format t "~a~%" fun))))
 
-(defun common-prefix (items)
-  (let ((lst 0))
-    (loop for n from 1 below (reduce #'min (mapcar #'length items)) do
-      (when (every (lambda (x)
-                      (char= (char (car items) n)
-                             (char x           n)))
-              (cdr items))
-       (setf lst n)))
-   (subseq (car items) 0 (1+ lst))))
-
-(defun starts-with (text)
-  (lambda (sym)
-    (let* ((symstr (string-downcase sym))
-           (cmp (subseq symstr 0 (min (length symstr) (length text)))))
-      (string= text cmp))))
-
-(defun select-completions (text list)
- (let* ((els (remove-if-not (starts-with text)
-                           (mapcar #'string list)))
-        (els (if (cdr els) (cons (common-prefix els) els) els)))
-    (if (string= text (string-downcase text))
-      (mapcar #'string-downcase els)
-      els)))
-
-(defun get-all-symbols ()
-  (let ((lst ()))
-    (do-all-symbols (s lst)
-      (when (or (fboundp s) (boundp s)) (push s lst)))
-    lst))
-
-(defun custom-complete (text start end)
-  (declare (ignore start) (ignore end))
-  (select-completions text (get-all-symbols)))
+(defun get-package-for-search (text)
+  "Return a list with:
+  - the text after the colon or double colon
+  - the package name (a string)
+  - T if we are looking for an external symbol, NIL for an internal one."
+  (let ((pos))
+    (cond
+      ((setf pos (search "::" text))
+       (list (subseq text  (+ pos 2))
+             (subseq text 0 pos)
+             nil))
+      ((setf pos (position #\: text))
+       (if (zerop pos)
+           (list text nil t)
+           (list (subseq text (1+ pos))
+                 (subseq text 0 pos)
+                 t)))
+      (t (list text nil  t)))))
+
+#+or(nil)
+(progn
+  (assert (equal (list "file-" "uiop" t)
+                 (get-package-for-search "uiop:file-")))
+  (assert (equal (list "*wil" "uiop" nil)
+                 (get-package-for-search "uiop::*wil")))
+  (assert (equal (list "*feat" nil t)
+                 (get-package-for-search "*feat"))))
+
+(defun list-external-symbols (sym-name pkg-name)
+  "List external symbols of PKG-NAME (a string).
+  (the symbol name is currently ignored)."
+  (declare (ignorable sym-name))
+  (assert (stringp pkg-name))
+  (loop :for sym :being :the :external-symbols :of (find-package pkg-name)
+     :collect (format nil "~(~a:~a~)" pkg-name sym)))
+
+(defun list-internal-symbols (sym-name pkg-name)
+  "List internal symbols of the package named PKG-NAME (a string)."
+  (declare (ignorable sym-name))
+  (assert (stringp pkg-name))
+  (loop :for sym :being :the :symbols :of (find-package pkg-name)
+     :collect (format nil "~(~a::~a~)" pkg-name sym)))
```

Comment:
Somehow the repetition here bugged me, so I came up with this replacement:

```common-lisp

(defmacro defsymlist (name docs selector s)
  `(defun ,name (sym-name pkg-name) ,docs
    (declare (ignorable sym-name))
    (assert (stringp pkg-name))
    (loop :for sym :being :the ,selector :of (find-package pkg-name)
      :collect (format nil ,s pkg-name sym))))

(defsymlist list-external-symbols
  "List external symbols of PKG-NAME (a string).
  (the symbol name is currently ignored)."
  :external-symbols "~(~a:~a~)")

(defsymlist list-internal-symbols
  "List internal symbols of the package named PKG-NAME (a string)."
  :symbols "~(~a::~a~)")
```

I understand that this is highly opinionated and macros like these are not for everyone, so I’ll leave it as a suggestion.

Reply:
mmh I don't find it worth it for these two little functions, so I leave this up to you :)

Reply:
I don’t really need this for peace of mind. Should we ever add another list, this might be worth revisiting.